R/shoji-catalog.R

```diff
@@ -240,28 +240,61 @@ setMethod("emails", "ShojiCatalog", function (x) getIndexSlot(x, "email"))
 #' @export
 as.list.ShojiCatalog <- function (x, ...) lapply(names(index(x)), function (i) x[[i]])
 
-#' Utility to get a more human-readable view of a Shoji Catalog
+#' A utility to retun a dataframe from a ShojiCatalog.
+#' 
+#' Some of the attributes of a ShojiCatalog will not naturally fit in
+#' a conventional dataframe. For instance a single variable might have multiple subvariables, 
+#' and these subvariables will not fit in a single row of a dataframe. In this case the
+#' list of subvariables are stored in a list-column in the resulting dataframe. 
 #'
 #' @param x ShojiCatalog or subclass
 #' @param keys character vector of attribute names from each catalog tuple to
 #' include in the result. Default is TRUE, which means all.
-#' @param rownames See \code{\link[base]{data.frame}}, the \code{row.names}
-#' argument, to which this is passed in \code{data.frame}. The difference here
-#' is that if \code{rownames} is explicitly set as \code{NULL}, the resulting
-#' object will not have row names set. By default, row names will be the URLs
-#' of the catalog tuples.
+#' @param rownames The rownames of the resulting dataframe. If set to NULL rownames 
+#' will default to the resulting dataframe will not have row names. By default the row names 
+#' will be the URLs of the catalog tuples.
+#' @param list_columns A character vector of the names of the attributes which should be stored
+#' in a list-column. 
 #' @param ... additional arguments passed to \code{data.frame}
 #' @return a \code{data.frame} view of the catalog
+#' @importFrom purrr map map_df map_lgl
+#' @importFrom tibble as_tibble
 #' @export
-catalogToDataFrame <- function (x, keys=TRUE, rownames, ...) {
+catalogToDataFrame <- function(x, keys=TRUE, 
+                                rownames = "default", 
+                                list_columns = c("subvariables", "subvariables_catalog"),
```

Comment:
I couldn't figure out a good way of doing that. The issue is that if I do something like ask whether `length(x) == length(unlist(x)` then I would wrongly thing that things like subvariables with only one subvariable should be stored as vectors. It also seems like we should make clear decisions about how each attribute should be stored, so coding the attribute name makes sense in that way. That said if anyone can think of a way of reliably inferring the proper output type that might also be a good solution.